src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,326 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// This file contains the protocol buffer representation of a list of supported
+// flags for Bazel commands.
+syntax = "proto3";
+
+package bazel_output_service;
+
+import "google/protobuf/any.proto";
+
+option java_package = "com.google.devtools.build.lib.remote";
+option java_outer_classname = "BazelOutputServiceProto";
+option go_package = "bazeloutputservice";
+
+// The Bazel Output Service may be used by users of the Remote Execution API to
+// construct a directory on the local system that contains all output files of a
+// build.
+//
+// Primitive implementations of this API may simply download files from the
+// Content Addressable Storage (CAS) and store them at their designated
+// location. Complex implementations may use a pseudo file system (e.g., FUSE)
+// to support deduplication, lazy loading and snapshotting.
+//
+// Details:
+// https://github.com/bazelbuild/proposals/blob/master/designs/2021-02-09-remote-output-service.md
+// https://groups.google.com/g/remote-execution-apis/c/qOSWWwBLPzo
+// https://groups.google.com/g/bazel-dev/c/lKzENsNd1Do
+// https://docs.google.com/document/d/1W6Tqq8cndssnDI0yzFSoj95oezRKcIhU57nwLHaN1qk/edit
+service BazelOutputService {
+  // Clean all data associated with a single output path, so that the
+  // next invocation of StartBuild() yields an empty output path. This
+  // may be implemented in a way that's faster than removing all of the
+  // files from the file system manually.
+  rpc Clean(CleanRequest) returns (CleanResponse);
+
+  // Signal that a new build is about to start.
+  //
+  // Bazel uses this call to obtain a directory where outputs of the build may
+  // be stored, called the output path. Based on the parameters provided, server
+  // may provide an empty output path, or one that has contents from a previous
+  // build of the same workspace.
+  //
+  // In case the output path contains data from a previous build, server is
+  // responsible for calling ContentAddressableStorage.FindMissingBlobs() for
+  // all of the objects that are stored remotely. This ensures that these
+  // objects don't disappear from the Content Addressable Storage while the
+  // build is running. Any files that are absent must be removed from the output
+  // path and reported through InitialOutputPathContents.modified_paths.
+  rpc StartBuild(StartBuildRequest) returns (StartBuildResponse);
+
+  // Stage build artifacts at the given paths with digests that are known to the
+  // Content Addressable Storage. Contents of these files can be lazily staged.
+  //
+  // If any of the digests or paths are invalid an error is returned without
+  // modifying anything. If some digests are unavailable, the corresponding
+  // StageArtifactsResponse.Response.unavailable flag should be set.
```

Comment:
So this is only describing the cases where digests/paths are invalid. This isn't actually covering the case where an object is not present in the CAS, right? Because I think the burden should be on Bazel to ensure it only calls StageArtifacts() when it knows the CAS objects exist. Otherwise the remote output service has to do redundant work.

If that's the case, should we maybe rename 'unavailable' to something else? Otherwise it could be misinterpreted as meaning 'CAS object not found'.

Reply:
`unavailable` indeed means `CAS object not found`. Bazel can ensure it only calls `StageArtifacts` on known CAS objects (got from cache lookup or execution result). But the object might be evicted before Bazel calls `StageArtifacts` (depending on how the CAS was configured).

`unavailable` is used to distinguish this case from other errors so that Bazel can probably rewind the build (or rewind the action if we ever implement it).

Reply:
Why couldn't Bazel call FindMissingBlobs to figure that out itself?

In the common case Bazel will call StageArtifacts in response to receiving an ActionResult. In that case it's guaranteed that objects in the CAS exist.

Reply:
> Why couldn't Bazel call FindMissingBlobs to figure that out itself?

Bazel could, but it needs to know when. In common case, when Bazel receives an `ActionResult`, it assumes the objects exist in CAS. Bazel won't proactively call `FindMissingBlobs` before or after calling `StageArtifacts`.

For servers that can guarantee objects in the CAS exist, they can simply ignore the `unavailable` field. For other servers that cannot make this guarantee due to reasons, they have a way to report back to Bazel.

Reply:
Got it. That sounds good. Buildbarn's storage is robust enough in that sense, so I can just let bb_clientd do nothing.

That said, instead of having a simple boolean field named `unavailable`, what are your thoughts on using a `google.rpc.Status` here? That way we can return `INVALID_ARGUMENT` in case the digest is invalid (e.g., not having the right hash length) and `NOT_FOUND` if the object is not present in the CAS.

Reply:
LGTM. Updated the proto.